slapo/schedule.py

```diff
@@ -1154,6 +1160,18 @@ def init_target_engine(sch, target, **kwargs):
     )
 
 
+def reduce_model_grads(sch: Schedule):
+    def hook(grad):
+        dist.all_reduce(grad.contiguous(), dist.ReduceOp.SUM, group=sch.group)
```

Comment:
This is not an efficient way to allreduce bias, since there are hundreds of bias in a large model. and we will end up with large communication latency, greatly slowing down the training. The best way for bias will be merging them in the end to perform a single allreduce call. for example, take a look at where layernorm gradient is reduced in megatron https://github.com/NVIDIA/Megatron-LM/blob/master/megatron/optimizer/optimizer.py#L285

Reply:
Right, I've seen the Megatron implementation, but the problem is that it is implemented in optimizer, so it can all_reduce the parameters in one place. For our case, we only return a model for users, and the computation and communication should be included in the model, so that's why I use param hooks instead.

Reply:
you can invoke gradient sync after the whole stage backward. nonetheless, we cannot use the current way of synchronizing gradients otherwise we will end up with extremely slow training example.